### PR TITLE
Update Kotlin language compatibility and stdlib to 1.8

### PR DESCRIPTION
**Pull request by bidetofevil**
## Goal

Update Kotlin language and stdlib compat to 1.8 so that a newer minimum version can be used in production. This enables us to update a tonne of other libraries

## Testing
Main tests for this is in the swazzler-test repo and there will be changes to that to ensure the right versions are tested

### CODE REVIEW

**Comment by bidetofevil** [comment]
* **#1057** <a href="https://app.graphite.dev/github/pr/embrace-io/embrace-android-sdk/1057?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @bidetofevil and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: buildSrc/src/main/kotlin/embrace-defaults.gradle.kts [hunk 1]
```diff
@@ -144,14 +144,10 @@ project.tasks.withType(JavaCompile::class.java).configureEach {
...
-        apiVersion = "1.4"
-        languageVersion = "1.4"
+        apiVersion = "1.7"
+        languageVersion = "1.7"
         jvmTarget = JavaVersion.VERSION_1_8.toString()
         freeCompilerArgs = freeCompilerArgs + "-Xexplicit-api=strict"
-
-        // Targetting Kotlin 1.4 emits a warning that can't be suppressed.
-        // Disabling this check for now.
-        allWarningsAsErrors = false
```
Comment (fractalwrench) [inline buildSrc/src/main/kotlin/embrace-defaults.gradle.kts:154]: 🎉

File: buildSrc/build.gradle.kts [hunk 2]
```diff
@@ -13,10 +13,12 @@ repositories {
 dependencies {
     compileOnly(gradleApi())
 
+    // Version of Kotlin used at build time
+    implementation("org.jetbrains.kotlin:kotlin-gradle-plugin:1.9.24")
```
Comment (bidetofevil) [inline buildSrc/build.gradle.kts:17]: part of me wants to go right into 2.0.... but I don't mind being a step behind

File: gradle/libs.versions.toml [hunk 3]
```diff
@@ -2,24 +2,19 @@
...
 openTelementrySemConv = "1.25.0-alpha"
-moshi = "1.12.0"
-lifecycle = "2.5.0"
-lifecycle-extensions = "2.0.0"
-mockk = "1.13.11"
-androidx-test = "1.6.1"
-androidx-test-junit = "1.2.1"
-robolectric = "4.13"
-mockwebserver = "4.12.0"
+moshi = "1.14.0"
```
Comment (bidetofevil) [inline gradle/libs.versions.toml:10]: This change doesn't move the Moshi codegen to use KSP. That will come in a different PR.

File: buildSrc/build.gradle.kts [hunk 4]
```diff
@@ -30,6 +32,8 @@ project.tasks.withType(JavaCompile::class.java).configureEach {
 // ensure the Kotlin + Java compilers both use the same language level.
 project.tasks.withType(KotlinCompile::class.java).configureEach {
     kotlinOptions {
-        jvmTarget = "1.8"
+        apiVersion = "1.7"
+        languageVersion = "1.7"
+        jvmTarget = JavaVersion.VERSION_1_8.toString()
```
Comment (fractalwrench) [inline buildSrc/build.gradle.kts:37]: I don't think this is strictly necessary as it applies to tasks in the build itself, although we could keep it if it doesn't cause any problems

File: gradle/libs.versions.toml [hunk 5]
```diff
@@ -2,24 +2,19 @@
...
+lifecycle = "2.8.3"
+lifecycle-extensions = "2.2.0"
 protobuf = "4.27.2"
-profileinstaller = "1.0.0"
-okhttp = "4.9.3"
+profileinstaller = "1.3.1"
+okhttp = "4.11.0"
 firebase = "23.1.0"
-compose = "1.0.5"
+compose = "1.5.4"
```
Comment (fractalwrench) [inline gradle/libs.versions.toml:17]: Is it essential to migrate to these deps as part of 1.7, and if so how can we test the changes for code that isn't really covered (such as compose listeners, unless I've missed the test coverage)?

**Review by fractalwrench — COMMENTED** [review]
There seems to be a CI failure around Kotlin toolchains not being available

File: gradle/libs.versions.toml (hunk 5 above)
Comment (bidetofevil) [reply gradle/libs.versions.toml:17]: I assumed there was test coverage, but let me revert those that don't. I think Lifecycle and OkHttp should be fine, but profile installer and compose may not be fully covered.

File: buildSrc/build.gradle.kts (hunk 4 above)
Comment (bidetofevil) [reply buildSrc/build.gradle.kts:37]: If it's not necessary I don't really want to add extra changes. I'll remove it

**Review by fractalwrench — APPROVED** [review]
LGTM

**Review by bidetofevil — COMMENTED** [review]
The `androidx.lifecycle` libraries are actually used by the Compose module, and should be upgraded in conjunction with the Compose version, which we've explicitly opted out of upgrading. As such, I'm revert this part of the upgrade for now.

**Pull request merged**